Upload the tutorial for the FastAPI search bot (#8314)
edgedb/edgedb: docs/intro/fastapi_gelai_searchbot.rst

        query: str

    class SearchResult(BaseModel):
        response: str | None = None
        sources: list[str] | None = None

Now we can define our endpoint and set the two classes we just added as its
argument and return type.

.. code-block:: python
    :caption: app/main.py

    @app.post("/search")
    async def search(search_terms: SearchTerms) -> SearchResult:
        return SearchResult(response=search_terms.query)

Same as before, we can test the endpoint using the UI, or by sending a request
with ``curl``:

.. code-block:: bash

   $ curl -X 'POST' \
      'http://127.0.0.1:8000/search' \
      -H 'accept: application/json' \
      -H 'Content-Type: application/json' \
      -d '{ "query": "string" }'

    {
      "response": "string",
      "sources": null
    }

3. Implement web search
=======================

Now that we have our web app infrastructure in place, let's add some substance
to it by implementing web search capabilities.

There're many powerful feature-rich products for LLM-driven web search. But for
purely educational purposes in this tutorial we'll be sailing on the high seas
🏴‍☠️and scraping Google search results directly. Google tends to actively

> resist such behavior, so the most reliable way for us to get our search results

Reviewer: Ouch ouch. Is this even within Google ToS? We can't really do this if they prohibit scraping them.

Reviewer: Maybe we do HN search instead? https://hn.algolia.com/api

Reviewer: That way:

* we're still making a real app
* we're not teaching people questionable things
* we don't have to deal with the beautifulsup garbage

Author: My justification is here is "no harm no foul". But sure, HN is a great idea!

Author: I will go with HN for now, but let me know if you think Brave'd be a better option despite token fiddling.

.. _ref_guide_fastapi_gelai_searchbot:

===================
FastAPI (Searchbot)
===================

> :edb-alt-title: Building a searchbot with memory using FastAPI and Gel AI

Reviewer 2: In this context (_a_ search bot), search bot should be two words

```suggestion
:edb-alt-title: Building a search bot with memory using FastAPI and Gel AI
```

In this tutorial we're going to walk you through building a chat bot with search
capabilities using Gel and `FastAPI <https://fastapi.tiangolo.com/>`_.

> FastAPI is a framework designed to help you build a web app, well, fast. And Gel
> is a data layer designed to help you figure out storage in your application, and
> also do it fast. By the end of this tutorial you will have tried out different
> aspects of using those two together.

Reviewer 2: ```suggestion
FastAPI is a framework designed to help you build web apps *fast*. Gel is a 
data layer designed to help you figure out storage in your application - also *fast*. 
By the end of this tutorial, you will have tried out different aspects of using those two together.
```

> We're going to start by creating an app with FastAPI, then add web search
> capabilities to it, and then put search results through a language model to get
> a human-friendly answer. After that we'll use Gel to implement chat history, and
> finish it off with semantic search, so that the bot can remember previous
> interactions with the user.

Reviewer 2: Wdyt? I changed the end so that "the 
bot remembers previous interactions with the user" refers to the chat history and not only the semantic search.
```suggestion
We will start by creating an app with FastAPI, adding web search
capabilities, and then putting search results through a language model to get
a human-friendly answer. After that, we'll use Gel to implement chat history so that the 
bot remembers previous interactions with the user. We'll finish it off with a semantic search. 
```

Author: Hmm, we're trying to explain that chat history enables multi-turn conversations, and semantic search adds cross-conversation "memory".

The end result is going to look something like this:

.. image::
    /docs/tutorials/placeholder.png
    :alt: Placeholder
    :width: 100%

1. Initialize the project
=========================

We're going to start by installing `uv <https://docs.astral.sh/uv/>`_ - a Python
package manager that's going to simplify environment management for us. You can
follow their `installation instructions
<https://docs.astral.sh/uv/getting-started/installation/>`_ or simply run:

.. code-block:: bash

    $ curl -LsSf https://astral.sh/uv/install.sh | sh

Once that is done, we can use uv to create scaffolding for our project following
the `documentation <https://docs.astral.sh/uv/guides/projects/>`_:

.. code-block:: bash

    $ uv init searchbot \
      && cd searchbot

> For now, we know we're going to need Gel and FastAPI, so let's add those
> following uv's instructions on `managing dependencies
> <https://docs.astral.sh/uv/concepts/projects/dependencies/#optional-dependencies>`_,
> as well as FastAPI's `installation docs
> <https://fastapi.tiangolo.com/#installation>`_. Running ``uv sync`` after that
> will create our virtual environment in a ``.venv`` directory and ensure it's
> ready. As a last step, we'll activate the environment and get started.

Reviewer 2: ```suggestion
For now, we know we're going to need Gel and FastAPI, so let's add those
following uv's instructions on `managing dependencies
<https://docs.astral.sh/uv/concepts/projects/dependencies/#optional-dependencies>`_,
as well as FastAPI's `installation docs
<https://fastapi.tiangolo.com/#installation>`_. Running ``uv sync`` after that
will create our virtual environment in a ``.venv`` directory and ensure it's
ready. As the last step, we'll activate the environment and get started.
```

.. code-block:: bash

    $ uv add "fastapi[standard]" \
      && uv add gel \
      && uv sync \
      && source .venv/bin/activate

.. note::

>     Make sure to source the environment every time you open a new terminal
>     session before running ``python``, ``gel`` or ``fastapi``-related commands.

Reviewer 2: ```suggestion
    Every time you open a new terminal session, you should source the environment
    before running ``python``, ``gel`` or ``fastapi`` commands.
```

2. Get started with FastAPI
===========================

At this stage we need to follow FastAPI's `tutorial
<https://fastapi.tiangolo.com/tutorial/>`_ to create the foundation of our app.

> We're going to make a super simple web API with one endpoint that takes in a

Reviewer 2: In general, I'd avoid "simple", "easy", etc, as it can be discouraging for beginners (easy for us != easy for other people). 

```suggestion
We're going to make a minimal web API with one endpoint that takes in a
```

user query as an input and echoes it as an output. First, let's make a directory
called ``app`` in our project root, and put an empty ``__init__.py`` there.

.. code-block:: bash

   $ mkdir app && touch app/__init__.py

Now let's create a file called ``main.py`` inside the ``app`` directory and put
the "Hello World" example in it:

.. code-block:: python
    :caption: app/main.py

    from fastapi import FastAPI

    app = FastAPI()

    @app.get("/")
    async def root():
        return {"message": "Hello World"}

To start the server, we'll run:

.. code-block:: bash

    $ fastapi dev app/main.py

> Once the server gets up and running, we can make sure it works using FastAPI's
> built-in UI at <http://127.0.0.1:8000/docs>_, or manually by using ``curl``:

Reviewer 2: ```suggestion
Once the server gets up and running, we can make sure it works using FastAPI's
built-in UI at <http://127.0.0.1:8000/docs>_, or manually with ``curl``:
```

.. code-block:: bash

    $ curl -X 'GET' \
      'http://127.0.0.1:8000/' \
      -H 'accept: application/json'

    {"message":"Hello World"}

> Now, in order to create the endpoint we set out to create, we need to pass our
> query as a parameter to it. We'd prefer to have it in the body of the request
> since user messages can get pretty long.

Reviewer 2: ```suggestion
Now, to create the search endpoint we mentioned earlier, we need to pass our
query as a parameter to it. We'd prefer to have it in the request's body
since user messages can be long.
```

> In FastAPI land this is done by creating a Pydantic schema and making it the
> type of the input parameter. `Pydantic <https://docs.pydantic.dev/latest/>`_ is
> a data validation library for Python. It has many features, but we don't
> actually need to know about them for now. All we need to know is that FastAPI
> uses Pydantic types to automatically figure out schemas for `input
> <https://fastapi.tiangolo.com/tutorial/body/>`_, as well as `output
> <https://fastapi.tiangolo.com/tutorial/response-model/>`_.

Reviewer 2: ```suggestion
In FastAPI land, this is done by creating a Pydantic schema and making it the
type of the input parameter. `Pydantic <https://docs.pydantic.dev/latest/>`_ is
a data validation library for Python. It has many features, but we don't
actually need to know about them for now. All we need to know is that FastAPI
uses Pydantic types to automatically figure out schemas for `input
<https://fastapi.tiangolo.com/tutorial/body/>`_, as well as `output
<https://fastapi.tiangolo.com/tutorial/response-model/>`_.
```

Let's add the following to our ``main.py``:

.. code-block:: python
    :caption: app/main.py

    from pydantic import BaseModel

    class SearchTerms(BaseModel):
        query: str

    class SearchResult(BaseModel):
        response: str | None = None

> Now we can define our endpoint and set the two classes we just added as its
> argument and return type.

Reviewer 2: ```suggestion
Now, we can define our endpoint. We'll set the two classes we just created as the new endpoint's
argument and return type.
```

.. code-block:: python
    :caption: app/main.py

    @app.post("/search")
    async def search(search_terms: SearchTerms) -> SearchResult:
        return SearchResult(response=search_terms.query)

Same as before, we can test the endpoint using the UI, or by sending a request
with ``curl``:

.. code-block:: bash

   $ curl -X 'POST' \
      'http://127.0.0.1:8000/search' \
      -H 'accept: application/json' \
      -H 'Content-Type: application/json' \
      -d '{ "query": "string" }'

    {
      "response": "string",
    }

3. Implement web search
=======================

Now that we have our web app infrastructure in place, let's add some substance
to it by implementing web search capabilities.

There're many powerful feature-rich products for LLM-driven web search. But in
this tutorial we're going to use a much more reliable source of real-world
information that is comment threads on Hacker News. Their web API is free of
charge and doesn't require an account. Below is a simple function that requests
a full-text search for a string query and extracts a nice sampling of comment
threads from each of the stories that came up in the result.

.. note::

    Link to HN

> We are not going to cover this code sample in too much depth. Feel free to grab

Author: I was worried about diluting the focus of the tutorial here. I think adding comments in the code example is a good compromise.

            response = requests.get(item_url)
            response.raise_for_status()
            item_result = response.json()

            site_url = f"https://news.ycombinator.com/item?id={hit['story_id']}"
            title = hit["title"]
            comments = extract_comment_thread(item_result)
            text = "\n".join(comments) if len(comments) > 0 else None
            web_sources.append(
                WebSource(url=site_url, title=title, text=text)
            )

        return web_sources

    if __name__ == "__main__":
        web_sources = fetch_web_sources("edgedb", limit=5)

        for source in web_sources:
            print(source.url)
            print(source.title)
            print(source.text)

Notice that we've created another Pydantic type called ``WebSource`` to store
our web search results. There's no framework-related reason for that, it's just
nicer than passing dictionaries around.

One more note: this snippet comes with an extra dependency called ``requests``,
which is a library for making HTTP requests. Let's add it by running:

.. code-block:: bash

    $ uv add requests

> Now we can test our web search on its own by running it like this:

Reviewer 2: ```suggestion
Now, we can test our web search on its own by running it like this:
```

.. code-block:: bash

    $ python3 app/web.py

It's time to reflect the new capabilities in our web app. We're going to hack

> our way around the fact that many stories on HN come with no comment threads.

Reviewer 2: Maybe let's explain what we're doing so that no one is scared by "hacking around" and expecting something tricky

.. code-block:: python
     :caption: app/main.py

     from .web import fetch_web_sources, WebSource

     async def search_web(query: str) -> list[WebSource]:
         raw_souces = fetch_web_sources(query, limit=30)
         return [s for s in raw_souces if s.text is not None][:5]

Now we can update the ``/search`` endpoint as follows:

.. code-block:: python-diff
    :caption: app/main.py

      class SearchResult(BaseModel):
          response: str | None = None
    +     sources: list[WebSource] | None = None

      @app.post("/search")
      async def search(search_terms: SearchTerms) -> SearchResult:
    +     web_sources = await search_web(search_terms.query)
    -     return SearchResult(response=search_terms.query)
    +     return SearchResult(
    +         response=search_terms.query, sources=web_sources
    +     )

4. Connect to the LLM
=====================

Now that we're capable of scraping text from search results, we can forward
those results to the LLM to get a nice-looking summary.

There's a million different LLMs accessible via a web API, feel free to choose

> whichever you prefer. In this tutorial we will roll with OpenAI, primarily for

Reviewer 2: People not that familiar with LLMs or building with LLMs won't know it - we can list a few of them instead

OpenAI's `API documentation
<https://platform.openai.com/docs/api-reference/chat/create>`_, and set up LLM
generation like this:

.. code-block:: python
    :caption: app/main.py

    import requests
    from dotenv import load_dotenv

    _ = load_dotenv()

    def get_llm_completion(system_prompt: str, messages: list[dict[str, str]]) -> str:
        api_key = os.getenv("OPENAI_API_KEY")
        url = "https://api.openai.com/v1/chat/completions"
        headers = {"Content-Type": "application/json", "Authorization": f"Bearer {api_key}"}

        response = requests.post(
            url,
            headers=headers,
            json={
                "model": "gpt-4o-mini",
                "messages": [
                    {"role": "developer", "content": system_prompt},
                    *messages,
                ],
            },
        )
        response.raise_for_status()
        result = response.json()
        return result["choices"][0]["message"]["content"]

Note that this cloud LLM API (and many others) requires a secret key to be set
as an environment variable. A common way to manage those is to use the

> ``python-dotenv`` library in combinations with a ``.env`` file. Feel free to
> browse `the readme
> <https://github.com/theskumar/python-dotenv?tab=readme-ov-file#getting-started>`_,
> although it's also quite simple to use. Create a file called ``.env`` in the

Reviewer 2: (So that we don't assume something is simple)
```suggestion
``python-dotenv`` library in combinations with a ``.env`` file. Feel free to
browse `the readme
<https://github.com/theskumar/python-dotenv?tab=readme-ov-file#getting-started>`_,
to learn more. Create a file called ``.env`` in the
```

root directory and put your api key in there:

.. code-block:: .env
    :caption: .env

    OPENAI_API_KEY="sk-..."

Don't forget to add the new dependency to the environment:

.. code-snippet:: bash

    uv add python-dotenv

And now we can integrate this LLM-related code with the rest of the app. First,

> let's set up the function that prepares LLM inputs:

Reviewer 2: ```suggestion
let's set up a function that prepares LLM inputs:
```

.. code-block:: python
    :caption: app/main.py

    async def generate_answer(
        query: str,
        web_sources: list[WebSource],
    ) -> SearchResult:
        system_prompt = (
            "You are a helpful assistant that answers user's questions"

>             + " by finding relevant information in HackerNews threads."

Reviewer 2: ```suggestion
            + " by finding relevant information in Hacker News threads."
```

    +     search_result = await generate_answer(search_terms.query, web_sources)
    +     return search_result
    -     return SearchResult(
    -         response=search_terms.query, sources=web_sources
    -     )

And now we can test the result as usual.

.. code-block:: bash

    $ curl -X 'POST' \
        'http://127.0.0.1:8000/search' \
        -H 'accept: application/json' \
        -H 'Content-Type: application/json' \
        -d '{ "query": "gel" }'

5. Use Gel to implement chat history
====================================

So far we've built an application that can take in a query, fetch some Hacker
News threads for it, sift through them using an LLM, and generate a nice
summary.

However, right now it's hardly user-friendly, since you have to speak in
keywords and basically start over every time you want to refine the query. To
enable more organic multi-turn interaction we need to add chat history and
infer the query from the context of the entire conversation.

Now's a good time to introduce Gel.

In case you need installation instructions, take a look at the :ref:`Quickstart
<ref_quickstart>`. Once Gel CLI is present in your system, initialize the

> project like this:

Reviewer 2: Now that's tricky. Since we'll have the quickstart as a more extended tutorial, should we have separate, short setup guides to refer people in cases like this one? 
cc @scotttrinh

Author: We can probably point to a reference manual instead. Unless there're specific non-trivial considerations or instructions there, in which case I think there needs to be a short guide, sure.

.. code-block:: bash

    $ gel project init --non-interactive

This command is going to put some project scaffolding inside our app, spin up a
local instace of Gel, and then link the two together. From now on, all
Gel-related things that happen inside our project directory are going to be
automatically run on the correct database instance, no need to worry about
connection incantations.

Defining the schema
-------------------

The database :ref:`schema <ref_datamodel_index>` in Gel is defined
declaratively. The :ref:`gel project init <ref_cli_edgedb_project_init>`
command has created a file called ``dbchema/default.esdl``, which we're going to
use to define our types.

We obviously want to keep track of messages, so we need to represent those in
the schema. By convention established in the LLM space, each message is going to
have a role in addition to the message content itself. We can also get Gel to
automatically keep track of message's creation time by adding a property callled

> ``timestamp`` and setting its :ref:`default value <ref_datamodel_props>` to the

Author: I guess it should. I'm honestly not super happy with the naming across the app right now - there're too many things named search-something, there's `search_terms.query` vs `search_query` (not the same thing) and so on. I'd like to rename everything in-bulk later, once all the major things are out of the way.
